server/src/main/java/org/elasticsearch/index/mapper/VersionEncoder.java

```diff
@@ -0,0 +1,256 @@
+                result.append(new BytesRef(part));
+            }
+
+            @Override
+            public DocValueFormat docValueFormat() {
+                return DocValueFormat.VERSION_SEMVER;
+            }
+
+        },
+        /**
+         * This mode will order mixed strings so that the numeric parts are treated with numeric ordering,
+         * e.g. "rc2" &lt; "rc11", "alpha523" &lt; "alpha1234"
+         */
+        HONOUR_NUMERALS
+        {
+            @Override
+            public void encode(String part, BytesRefBuilder result) {
+                prefixDigitGroupsWithLength(part, result);
+            }
+
+            @Override
+            public DocValueFormat docValueFormat() {
+                return DocValueFormat.VERSION_NUMERIC;
+            }
+        };
+
+        public abstract void encode(String part, BytesRefBuilder result);
+        public abstract DocValueFormat docValueFormat();
+
+        public static SortMode fromString(String mode) {
+            switch (mode) {
+                case "semver":
+                    return SEMVER;
+                case "numeric":
+                    return HONOUR_NUMERALS;
+                default:
+                    throw new IllegalArgumentException("Unknown version field mode: " + mode);
+            }
+        }
+
+    }
+
+    /**
+     * Encodes a version string given the given {@link SortMode}.
+     * First, the input version string is split into the following parts:
+     * <p>
+     * mainVersion(-preReleasePart)(+buildId)
+     *
+     */
+    public static BytesRef encodeVersion(String versionString, SortMode mode) {
+        //System.out.println("encoding: " + versionString);
+        if (legalVersionString(versionString) == false) {
+            throw new IllegalArgumentException("Illegal version string: " + versionString);
+        }
+        // extract "build" suffix starting with "+"
+        String buildSuffixPart = extractSuffix(versionString, BUILD_SEPARATOR);
+        if (buildSuffixPart != null) {
+           versionString = versionString.substring(0, versionString.length() - buildSuffixPart.length());
+        }
+
+        // extract "pre-release" suffix starting with "-"
+        String preReleaseId = extractSuffix(versionString, PRERELESE_SEPARATOR);
+        if (preReleaseId != null) {
+            versionString = versionString.substring(0, versionString.length() - preReleaseId.length());
+        }
+
+        // pad all digit groups in main part with numeric marker and length bytes
+        BytesRefBuilder encodedVersion = new BytesRefBuilder();
+        prefixDigitGroupsWithLength(versionString, encodedVersion);
+
+        // encode whether version has pre-release parts
+        if (preReleaseId != null) {
+            encodedVersion.append(PRERELESE_SEPARATOR_BYTE);  // versions with pre-release part sort before ones without
+            String[] preReleaseParts = preReleaseId.substring(1).split(DOT_SEPARATOR_REGEX);
```

Comment:
If we want to support version strings that have consecutive dots, then I think we should use another split method as this one would skip the empty string in such a case.

Reply:
I haven't seen reason for supporting consecutive dots yet, currently I would try to catch that in initial validation. Do you have something in mind that would require this?